libvast/vast/fbs/table.hpp

```diff
@@ -0,0 +1,171 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include "vast/chunk.hpp"
+#include "vast/detail/assert.hpp"
+#include "vast/fwd.hpp"
+
+#include <caf/error.hpp>
+#include <caf/meta/load_callback.hpp>
+#include <caf/meta/save_callback.hpp>
+#include <caf/meta/type_name.hpp>
+
+#include <flatbuffers/flatbuffers.h>
+
+namespace vast::fbs {
+
+/// Semi-regular base class for types that wrap a FlatBuffers table.
+/// @tparam Derived The type that wraps the FlatBuffers table for CRTP.
+/// @tparam Root The generated FlatBuffers table root type.
+/// @tparam FileIdentifier A pointer to the function that returns the
+/// FlatBuffers table's file identifier.
+template <class Derived, class Root, const char* (*FileIdentifier)()>
+class table {
+public:
+  // -- types and constants ----------------------------------------------------
+
+  using derived_type = Derived;
+  using root_type = Root;
+
+  /// A pointer to the function that returns the underlying FlatBuffers table's
+  /// file identifier.
+  inline static constexpr auto file_identifier = FileIdentifier;
+
+  /// Returns the underlying FlatBuffers table's fully qualified table name.
+  static constexpr const char* table_name() {
+    return root_type::GetFullyQualifiedName();
+  }
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  /// Default-constructs an invalid FlatBuffers table.
+  /// @post `!*this`
+  table() noexcept = default;
+
+  /// Constructs and verifies a FlatBuffers table from a chunk.
+  /// @note Constructs an invalid table if the chunk fails verification.
+  explicit table(chunk_ptr&& chunk) noexcept {
+    if (chunk) {
+      auto verifier = flatbuffers::Verifier{
+        reinterpret_cast<const uint8_t*>(chunk->data()), chunk->size()};
+      if (verifier.template VerifyBuffer<root_type>(file_identifier()))
```

Comment:
I'm not sure we always want to verify by default, at least for partitions we intentionally only access a subset of the full flatbuffer at startup.

Reply:
In a call we agreed to add an optional tag type parameter that disables the verification.